helpers/remoteoauth/token.go

```diff
@@ -0,0 +1,170 @@
+package remoteoauth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	cloudquery_api "github.com/cloudquery/cloudquery-api-go"
+	"github.com/google/uuid"
+	"golang.org/x/oauth2"
+)
+
+type Token struct {
+	currentToken oauth2.Token
+	remoteToken  bool         // flag to indicate if the current token was retrieved from a remote source
+	tokenMu      sync.RWMutex // protects the variables above
+
+	apiClient *cloudquery_api.ClientWithResponses
+	mu        sync.Mutex // protects against multiple refresh calls
+
+	cloudEnabled     bool
+	apiURL           string
+	apiToken         string
+	teamName         string
+	syncName         string
+	testConnUUID     uuid.UUID
+	syncRunUUID      uuid.UUID
+	connectorUUID    uuid.UUID
+	isTestConnection bool
+}
+
+var (
+	ErrTokenExpired = errors.New("token expired and cloud env is not set")
+
+	_ oauth2.TokenSource = (*Token)(nil)
+)
+
+// defaultExpiryDelta determines how earlier a token should be considered
+// expired than its actual expiration time. It is used to avoid late
+// expirations due to client-server time mismatches.
+const defaultExpiryDelta = 10 * time.Second
+
+func NewToken(opts ...TokenOption) (*Token, error) {
+	t := &Token{}
+	for _, opt := range opts {
+		opt(t)
+	}
+	err := t.initCloudOpts()
+	if err != nil {
+		return nil, err
+	}
+	if t.cloudEnabled {
+		t.apiClient, err = cloudquery_api.NewClientWithResponses(t.apiURL,
+			cloudquery_api.WithRequestEditorFn(func(_ context.Context, req *http.Request) error {
+				req.Header.Set("Authorization", "Bearer "+t.apiToken)
+				return nil
+			}))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create api client: %w", err)
+		}
+	}
+
+	return t, nil
+}
+
+// Token returns the cached token if not expired, or a new token from the remote source.
+func (t *Token) Token() (*oauth2.Token, error) {
+	return t.TokenWithContext(context.TODO())
```

Comment:
Is this something we should support? What happens if we want to cancel the context/request?

Reply:
This is the interface of `oauth2.TokenSource`. If we don't support this, this can't be swapped in for plugins which use the normal/standard oauth2 package methods.